letsencrypt/client/tests/ddns/ddns_test.py

```diff
@@ -0,0 +1,77 @@
+"""Test for letsencrypt.client.ddns."""
+import unittest
+
+import mock
+
+from letsencrypt.client import challenge_util
+
+
+class DDNSPerformTest(unittest.TestCase):
+    """Test the DNS challenge (for dynamic DNS zones)."""
+
+    def setUp(self):
+        super(DDNSPerformTest, self).setUp()
+
+        from letsencrypt.client.ddns import ddns
+        self.ddns = ddns.DDNS()
+
+        # XXX unclear: what is key?
+        key = None
+        self.challs = []
+        self.challs.append(challenge_util.DnsChall(
+            "encryption-example.demo", "17817c66b60ce2e4012dfad92657527a", key))
+        self.challs.append(challenge_util.DnsChall(
+            "letsencrypt.demo", "27817c66b60ce2e4012dfad92657527a", key))
+
+    def tearDown(self):
+        pass
+
+    def test_pref(self):
+        result = self.ddns.get_chall_pref("example.org")
+        self.assertEqual(result, ["dns"])
+
+    def test_perform0(self):
+        resp = self.ddns.perform([])
+        self.assertTrue(resp == [])
+
+    @mock.patch("letsencrypt.client.ddns.ddns.nsupdate")
+    def test_perform2(self, mock_nsupdate):
+        responses = self.ddns.perform(self.challs)
+
+        self.assertEqual(mock_nsupdate.call_count, 2)
+        calls = mock_nsupdate.call_args_list
+        expected_call_list = [
+            ("add", "encryption-example.demo",
+             "17817c66b60ce2e4012dfad92657527a", ),
+            ("add", "letsencrypt.demo",
+             "27817c66b60ce2e4012dfad92657527a", ),
+        ]
+
+        for i in range(len(expected_call_list)):
```

Comment:
there will be some compatibility library that chooses correctly xrange (py2) or range (py3), so this is a minor problem. supporting 2.6 (and py3 <3.3 also) are bigger problems.